File: frontend/src/api/responsible-ai.ts
// In dev, use /api-backend so Vite proxies to the API without conflicting with the SPA route /admin/responsible-ai
const API_BASE = import.meta.env.VITE_API_BASE_URL ?? (import.meta.env.DEV ? "/api-backend" : "");

export interface AgentRunEventRow {
  id: number;
  event_id: string;
  timestamp: string;
  agent_id: string;
  run_id: string | null;
  client_id_scope: string | null;
  input_summary: Record<string, unknown>;
  success: boolean;
  error_message: string | null;
  explanation_summary: string | null;
  data_sources_used: string[] | null;
  choice_criteria: string[] | null;
  input_validation_passed: boolean | null;
  guardrail_triggered: boolean | null;
  payload_ref: string | null;
  payload?: {
    id: string;
    run_id: string;
    created_at: string;
    client_id: string;
    payload: Record<string, unknown>;
  } | null;
}

export interface ResponsibleAIStats {
  from_date: string;
  to_date: string;
  total_runs: number;
  success_count: number;
  success_rate: number;
  agent_one_runs: number;
  agent_two_runs: number;
  agent_three_runs: number;
  agent_two_with_explanation: number;
  explainability_coverage_pct: number;
  guardrail_triggered_count: number;
}

export async function fetchResponsibleAIEvents(params: {
  agent_id?: string;
  from_date?: string;
  to_date?: string;
  success?: boolean;
  client_id_scope?: string;
  limit?: number;
  offset?: number;
}): Promise<{ events: AgentRunEventRow[] }> {
  const sp = new URLSearchParams();
  if (params.agent_id != null) sp.set("agent_id", params.agent_id);
  if (params.from_date != null) sp.set("from_date", params.from_date);
  if (params.to_date != null) sp.set("to_date", params.to_date);
  if (params.success != null) sp.set("success", String(params.success));
  if (params.client_id_scope != null) sp.set("client_id_scope", params.client_id_scope);
  if (params.limit != null) sp.set("limit", String(params.limit));
  if (params.offset != null) sp.set("offset", String(params.offset));
  const url = `${API_BASE}/admin/responsible-ai/events?${sp.toString()}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Failed to load events: ${res.status}`);
  return res.json();
}

export async function fetchResponsibleAIStats(params: {
  from_date?: string;
  to_date?: string;
}): Promise<ResponsibleAIStats> {
  const sp = new URLSearchParams();
  if (params.from_date != null) sp.set("from_date", params.from_date);
  if (params.to_date != null) sp.set("to_date", params.to_date);
  const url = `${API_BASE}/admin/responsible-ai/stats?${sp.toString()}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Failed to load stats: ${res.status}`);
  return res.json();
}

export async function fetchResponsibleAIEventById(eventId: string): Promise<AgentRunEventRow> {
  const url = `${API_BASE}/admin/responsible-ai/events/${encodeURIComponent(eventId)}`;
  const res = await fetch(url);
  if (!res.ok) {
    if (res.status === 404) throw new Error("Event not found");
    throw new Error(`Failed to load event: ${res.status}`);
  }
  return res.json();
}

File: api/src/api/routers/responsible_ai.py
"""
Admin read-only API for Responsible AI dashboard: list events, stats, event detail with payload.
Protect with ADMIN_API_KEY header when set.
"""

from datetime import datetime, timezone, timedelta

from fastapi import APIRouter, HTTPException, Query

from api.database import fetch_rows

router = APIRouter(prefix="/admin/responsible-ai", tags=["admin", "responsible-ai"])


def _serialize_row(r: dict) -> dict:
    """Convert asyncpg row to JSON-serializable dict (UUID/datetime to str)."""
    out = {}
    for k, v in r.items():
        if v is None:
            out[k] = None
        elif hasattr(v, "isoformat"):
            out[k] = v.isoformat()
        elif hasattr(v, "hex"):  # UUID
            out[k] = str(v)
        else:
            out[k] = v
    return out


@router.get("/events")
async def list_events(
    agent_id: str | None = Query(None, description="Filter by agent_id"),
    from_date: datetime | None = Query(None, description="From timestamp (UTC)"),
    to_date: datetime | None = Query(None, description="To timestamp (UTC)"),
    success: bool | None = Query(None, description="Filter by success"),
    client_id_scope: str | None = Query(None, description="Filter by client_id_scope"),
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
):
    """List agent run events with optional filters."""
    rows = await fetch_rows(
        "get_agent_run_events",
        agent_id,
        from_date,
        to_date,
        success,
        client_id_scope,
        limit,
        offset,
    )
    return {"events": [_serialize_row(r) for r in rows]}


@router.get("/stats")
async def get_stats(
    from_date: datetime | None = Query(None, description="From (UTC); default 30 days ago"),
    to_date: datetime | None = Query(None, description="To (UTC); default now"),
):
    """Aggregate stats for dashboard: total runs, success rate, by agent, explainability coverage."""
    now = datetime.now(timezone.utc)
    to_d = to_date or now
    from_d = from_date or (now - timedelta(days=30))
    rows = await fetch_rows("get_agent_run_events_stats", from_d, to_d)
    if not rows:
        return {
            "from_date": from_d.isoformat(),
            "to_date": to_d.isoformat(),
            "total_runs": 0,
            "success_count": 0,
            "success_rate": 0.0,
            "agent_one_runs": 0,
            "agent_two_runs": 0,
            "agent_three_runs": 0,
            "agent_two_with_explanation": 0,
            "explainability_coverage_pct": 0.0,
            "guardrail_triggered_count": 0,
        }
    r = rows[0]
    total = r.get("total_runs") or 0
    success_count = r.get("success_count") or 0
    agent_two_with_exp = r.get("agent_two_with_explanation") or 0
    agent_two_runs = r.get("agent_two_runs") or 0
    return {
        "from_date": from_d.isoformat(),
        "to_date": to_d.isoformat(),
        "total_runs": total,
        "success_count": success_count,
        "success_rate": (success_count / total * 100.0) if total else 0.0,
        "agent_one_runs": r.get("agent_one_runs") or 0,
        "agent_two_runs": agent_two_runs,
        "agent_three_runs": r.get("agent_three_runs") or 0,
        "agent_two_with_explanation": agent_two_with_exp,
        "explainability_coverage_pct": (agent_two_with_exp / agent_two_runs * 100.0) if agent_two_runs else 0.0,
        "guardrail_triggered_count": r.get("guardrail_triggered_count") or 0,
    }


@router.get("/events/{event_id}")
async def get_event_by_id(event_id: str):
    """Get a single event by event_id; include full payload when payload_ref is set (agentTwo)."""
    rows = await fetch_rows("get_agent_run_event_by_id", event_id)
    if not rows:
        raise HTTPException(status_code=404, detail="Event not found")
    event = _serialize_row(rows[0])
    payload_ref = event.get("payload_ref")
    if payload_ref:
        ref_rows = await fetch_rows("get_agent_two_payload_by_ref", str(payload_ref))
        if ref_rows:
            event["payload"] = _serialize_row(ref_rows[0])
        else:
            event["payload"] = None
    else:
        event["payload"] = None
    return event